modules/navier_stokes/src/functormaterials/INSFVEnthalpyFunctorMaterial.C

```diff
@@ -32,31 +33,92 @@ INSFVEnthalpyFunctorMaterial::validParams()
+      "Specific enthalpy functor, to be used when cp is not constant to compute the enthalpy, as "
+      "an alternative to using a 'fp' FluidProperties object");
+
   return params;
 }
 
 INSFVEnthalpyFunctorMaterial::INSFVEnthalpyFunctorMaterial(const InputParameters & parameters)
   : FunctorMaterial(parameters),
+    _assumed_constant_cp(getParam<bool>("assumed_constant_cp")),
+    _fp(isParamValid(NS::fluid)
+            ? &UserObjectInterface::getUserObject<SinglePhaseFluidProperties>(NS::fluid)
+            : nullptr),
     _rho(getFunctor<ADReal>(NS::density)),
     _temperature(getFunctor<ADReal>("temperature")),
-    _cp(getFunctor<ADReal>(NS::cp))
+    _pressure(isParamValid("pressure") ? &getFunctor<ADReal>("pressure") : nullptr),
+    _cp(getFunctor<ADReal>(NS::cp)),
+    _h(isParamValid(NS::specific_enthalpy + "_in")
+           ? &getFunctor<ADReal>(NS::specific_enthalpy + "_in")
+           : nullptr)
 {
-  const auto & rho_h =
-      addFunctorProperty<ADReal>(NS::enthalpy_density,
-                                 [this](const auto & r, const auto & t)
-                                 { return _rho(r, t) * _cp(r, t) * _temperature(r, t); });
+  // We have to use a warning because fp is often in the global parameters
+  if (_assumed_constant_cp && _fp)
+    paramWarning(
+        "fp", "No need to specify fluid properties if assuming the specific enthalpy is constant");
+  if (!_assumed_constant_cp && ((!_fp || !_pressure) && !_h))
+    paramError("fp",
+               "Must specify both fluid properties and pressure or an enthalpy functor if not "
+               "assuming the specific enthalpy is constant");
+
+  if (_assumed_constant_cp)
+  {
+    const auto & rho_h =
+        addFunctorProperty<ADReal>(NS::enthalpy_density,
+                                   [this](const auto & r, const auto & t)
+                                   { return _rho(r, t) * _cp(r, t) * _temperature(r, t); });
+
+    const auto & h = addFunctorProperty<ADReal>(NS::specific_enthalpy,
+                                                [this](const auto & r, const auto & t)
+                                                { return _cp(r, t) * _temperature(r, t); });
+
+    addFunctorProperty<ADReal>(NS::time_deriv(getParam<MooseFunctorName>(NS::specific_enthalpy)),
+                               [this](const auto & r, const auto & t)
+                               { return _cp(r, t) * _temperature.dot(r, t); });
+
+    addFunctorProperty<ADReal>(
+        "rho_cp_temp", [&rho_h](const auto & r, const auto & t) -> ADReal { return rho_h(r, t); });
 
-  const auto & h = addFunctorProperty<ADReal>(NS::specific_enthalpy,
-                                              [this](const auto & r, const auto & t)
-                                              { return _cp(r, t) * _temperature(r, t); });
+    addFunctorProperty<ADReal>("cp_temp",
+                               [&h](const auto & r, const auto & t) -> ADReal { return h(r, t); });
```

Comment:
Did these stay because you don't want to modify inputs/other source files that grab these under the hood? I think it would be nicer if we didn't have these and had something uniform.

Reply:
Sounds good!